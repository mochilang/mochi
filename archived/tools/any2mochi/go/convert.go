package golang

import (
	"bytes"
	"fmt"
	"go/ast"
	"go/printer"
	"go/scanner"
	"go/token"
	"os"
	"strings"
	"unicode"
)

// ConvertError provides a detailed error suitable for golden tests.
type ConvertError struct {
	Line   int
	Column int
	Msg    string
	Snip   string
}

var mochiReserved = map[string]bool{
	"let": true, "fun": true, "if": true, "else": true, "for": true,
	"return": true, "true": true, "false": true, "stream": true,
	"agent": true, "test": true, "expect": true, "on": true,
	"intent": true, "import": true, "extern": true, "var": true,
	"break": true, "continue": true, "while": true, "in": true,
	"generate": true, "match": true, "fetch": true, "load": true,
	"save": true, "package": true, "export": true, "fact": true,
	"rule": true, "all": true, "null": true,
}

func sanitizeName(name string) string {
	if name == "" {
		return ""
	}
	var b strings.Builder
	for i, r := range name {
		if unicode.IsLetter(r) || r == '_' || (unicode.IsDigit(r) && i > 0) {
			b.WriteRune(r)
		} else {
			b.WriteRune('_')
		}
	}
	out := b.String()
	if out == "" || !(unicode.IsLetter(rune(out[0])) || out[0] == '_') {
		out = "_" + out
	}
	if mochiReserved[out] {
		out = "_" + out
	}
	return out
}

func (e *ConvertError) Error() string {
	if e.Line > 0 {
		if e.Column > 0 {
			return fmt.Sprintf("line %d:%d: %s\n%s", e.Line, e.Column, e.Msg, e.Snip)
		}
		return fmt.Sprintf("line %d: %s\n%s", e.Line, e.Msg, e.Snip)
	}
	return fmt.Sprintf("%s\n%s", e.Msg, e.Snip)
}

// Convert converts Go source code to Mochi without relying on a
// language server. It parses the source using the standard library
// and emits minimal Mochi stubs.
func Convert(src string) ([]byte, error) {
	ast, err := ParseAST(src)
	if err != nil {
		return nil, formatParseError(src, err)
	}
	out := ConvertAST(ast)
	if len(out) == 0 {
		return nil, &ConvertError{Msg: "no convertible symbols found", Snip: numberedSnippet(src)}
	}
	return out, nil
}

// ConvertFile reads the Go file at path and converts it to Mochi.
func ConvertFile(path string) ([]byte, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	out, err := Convert(string(data))
	if err != nil {
		return nil, formatError(path, string(data), err)
	}
	return out, nil
}

func writeFunc(out *strings.Builder, fset *token.FileSet, fn *ast.FuncDecl) {
	if fn.Name == nil {
		return
	}
	name := sanitizeName(fn.Name.Name)
	if fn.Recv != nil && len(fn.Recv.List) > 0 {
		recv := exprString(fset, fn.Recv.List[0].Type)
		name = sanitizeName(recv) + "." + name
	}
	out.WriteString("fun ")
	out.WriteString(name)
	out.WriteByte('(')
	first := true
	if fn.Type.Params != nil {
		for _, p := range fn.Type.Params.List {
			typ := exprString(fset, p.Type)
			if len(p.Names) == 0 {
				if !first {
					out.WriteString(", ")
				}
				out.WriteString("_")
				if typ != "" {
					out.WriteString(": ")
					out.WriteString(typ)
				}
				first = false
				continue
			}
			for _, n := range p.Names {
				if !first {
					out.WriteString(", ")
				}
				out.WriteString(sanitizeName(n.Name))
				if typ != "" {
					out.WriteString(": ")
					out.WriteString(typ)
				}
				first = false
			}
		}
	}
	out.WriteByte(')')
	if fn.Type.Results != nil {
		var rets []string
		for _, r := range fn.Type.Results.List {
			typ := exprString(fset, r.Type)
			if typ == "" || typ == "void" {
				continue
			}
			n := len(r.Names)
			if n == 0 {
				n = 1
			}
			for i := 0; i < n; i++ {
				rets = append(rets, typ)
			}
		}
		if len(rets) == 1 {
			out.WriteString(": ")
			out.WriteString(rets[0])
		} else if len(rets) > 1 {
			out.WriteString(": (")
			out.WriteString(strings.Join(rets, ", "))
			out.WriteString(")")
		}
	}
	out.WriteString(" {}\n")
}

func writeGenDecl(out *strings.Builder, fset *token.FileSet, d *ast.GenDecl) {
	switch d.Tok {
	case token.TYPE:
		for _, spec := range d.Specs {
			ts, ok := spec.(*ast.TypeSpec)
			if !ok {
				continue
			}
			if st, ok := ts.Type.(*ast.StructType); ok {
				out.WriteString("type ")
				out.WriteString(sanitizeName(ts.Name.Name))
				out.WriteString(" {\n")
				for _, f := range st.Fields.List {
					typ := exprString(fset, f.Type)
					if len(f.Names) == 0 {
						out.WriteString("  ")
						out.WriteString(typ)
						out.WriteByte('\n')
						continue
					}
					for _, n := range f.Names {
						out.WriteString("  ")
						out.WriteString(sanitizeName(n.Name))
						if typ != "" {
							out.WriteString(": ")
							out.WriteString(typ)
						}
						out.WriteByte('\n')
					}
				}
				out.WriteString("}\n")
			}
		}
	case token.VAR, token.CONST:
		for _, spec := range d.Specs {
			vs, ok := spec.(*ast.ValueSpec)
			if !ok {
				continue
			}
			typ := ""
			if vs.Type != nil {
				typ = exprString(fset, vs.Type)
			}
			for _, n := range vs.Names {
				out.WriteString("let ")
				out.WriteString(sanitizeName(n.Name))
				if typ != "" {
					out.WriteString(": ")
					out.WriteString(typ)
				}
				out.WriteByte('\n')
			}
		}
	}
}

func exprString(fset *token.FileSet, e ast.Expr) string {
	var b bytes.Buffer
	printer.Fprint(&b, fset, e)
	return simplifyType(b.String())
}

func numberedSnippet(src string) string {
	lines := strings.Split(src, "\n")
	if len(lines) > 10 {
		lines = lines[:10]
	}
	for i, l := range lines {
		lines[i] = fmt.Sprintf("%3d: %s", i+1, l)
	}
	return strings.Join(lines, "\n")
}

func simplifyType(t string) string {
	t = strings.TrimSpace(t)
	if strings.HasPrefix(t, "*") {
		return simplifyType(strings.TrimPrefix(t, "*"))
	}
	if strings.HasPrefix(t, "[]") {
		return "list<" + simplifyType(t[2:]) + ">"
	}
	if strings.HasPrefix(t, "map[") {
		end := strings.Index(t, "]")
		if end > 0 {
			key := simplifyType(t[4:end])
			val := simplifyType(t[end+1:])
			return fmt.Sprintf("map<%s, %s>", key, val)
		}
	}
	if strings.HasPrefix(t, "func(") {
		end := strings.LastIndex(t, ")")
		if end > 0 {
			paramsPart := t[len("func("):end]
			rest := strings.TrimSpace(t[end+1:])
			var params []string
			if strings.TrimSpace(paramsPart) != "" {
				for _, p := range strings.Split(paramsPart, ",") {
					p = strings.TrimSpace(p)
					if strings.HasPrefix(p, "...") {
						t := simplifyType(strings.TrimSpace(p[3:]))
						params = append(params, "list<"+t+">")
					} else {
						params = append(params, simplifyType(p))
					}
				}
			}
			paramStr := strings.Join(params, ", ")
			if rest != "" {
				ret := strings.Trim(rest, "() ")
				ret = simplifyType(ret)
				if strings.Contains(rest, ",") {
					ret = "(" + ret + ")"
				}
				return fmt.Sprintf("fun(%s): %s", paramStr, ret)
			}
			return fmt.Sprintf("fun(%s)", paramStr)
		}
	}
	if t == "interface{}" {
		return "any"
	}
	// Replace package qualifiers in type names with underscores so that the
	// resulting identifier is a valid Mochi type reference. This avoids
	// parse errors for types like "time.Duration" which are emitted by the
	// Go compiler for helper functions.
	if strings.ContainsAny(t, "./") {
		t = strings.NewReplacer(".", "_", "/", "_").Replace(t)
	}
	return t
}

func formatParseError(src string, err error) error {
	if el, ok := err.(scanner.ErrorList); ok && len(el) > 0 {
		e := el[0]
		pos := e.Pos
		line := int(pos.Line)
		col := int(pos.Column)
		return &ConvertError{Line: line, Column: col, Msg: e.Msg, Snip: arrowSnippet(src, line, col)}
	}
	return err
}

func arrowSnippet(src string, line, col int) string {
	lines := strings.Split(src, "\n")
	start := line - 3
	if start < 0 {
		start = 0
	}
	end := line + 2
	if end > len(lines) {
		end = len(lines)
	}
	var b strings.Builder
	for i := start; i < end; i++ {
		mark := "   "
		if i+1 == line {
			mark = ">>>"
		}
		fmt.Fprintf(&b, "%4d:%s %s\n", i+1, mark, lines[i])
		if i+1 == line {
			b.WriteString("     " + strings.Repeat(" ", col-1) + "^\n")
		}
	}
	return strings.TrimRight(b.String(), "\n")
}

func formatError(path, src string, err error) error {
	if ce, ok := err.(*ConvertError); ok {
		if ce.Line > 0 {
			return fmt.Errorf("%s:%d:%d: %s\n%s", path, ce.Line, ce.Column, ce.Msg, ce.Snip)
		}
		return fmt.Errorf("%s: %s\n%s", path, ce.Msg, ce.Snip)
	}
	if el, ok := err.(scanner.ErrorList); ok && len(el) > 0 {
		e := el[0]
		line := int(e.Pos.Line)
		col := int(e.Pos.Column)
		return fmt.Errorf("%s:%d:%d: %s\n%s", path, line, col, e.Msg, arrowSnippet(src, line, col))
	}
	return fmt.Errorf("%s: %v", path, err)
}
