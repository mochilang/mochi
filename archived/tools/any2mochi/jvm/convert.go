//go:build archive

package jvm

import (
	"fmt"
	"os"
	"strings"

	parent "mochi/archived/tools/any2mochi"
)

// Convert converts JVM assembly source to Mochi using the language server.
func Convert(src string) ([]byte, error) {
	ls := parent.Servers["jvm"]
	syms, diags, err := parent.EnsureAndParse(ls.Command, ls.Args, ls.LangID, src)
	if err != nil {
		return nil, err
	}
	if len(diags) > 0 {
		return nil, fmt.Errorf("%s", formatDiagnostics(src, diags))
	}
	var out strings.Builder
	for _, s := range syms {
		if s.Kind != parent.SymbolKindFunction {
			continue
		}
		out.WriteString("fun ")
		out.WriteString(s.Name)
		out.WriteString("() {}\n")
	}
	if out.Len() == 0 {
		return nil, fmt.Errorf("no convertible symbols found\n\nsource snippet:\n%s", numberedSnippet(src))
	}
	return []byte(out.String()), nil
}

// ConvertFile reads the JVM file and converts it to Mochi.
func ConvertFile(path string) ([]byte, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return Convert(string(data))
}

func numberedSnippet(src string) string {
	lines := strings.Split(src, "\n")
	if len(lines) > 10 {
		lines = lines[:10]
	}
	for i, l := range lines {
		lines[i] = fmt.Sprintf("%3d: %s", i+1, l)
	}
	return strings.Join(lines, "\n")
}

func formatDiagnostics(src string, diags []parent.Diagnostic) string {
	lines := strings.Split(src, "\n")
	var out strings.Builder
	for _, d := range diags {
		start := int(d.Range.Start.Line)
		msg := d.Message
		line := ""
		if start < len(lines) {
			line = strings.TrimSpace(lines[start])
		}
		out.WriteString(fmt.Sprintf("line %d: %s\n  %s\n", start+1, msg, line))
	}
	return strings.TrimSpace(out.String())
}
