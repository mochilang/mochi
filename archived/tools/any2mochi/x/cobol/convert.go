//go:build archive

package cobol

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

// Convert converts COBOL source code to Mochi.
func Convert(src string) ([]byte, error) {
	ast, err := Parse(src)
	if err != nil {
		return nil, err
	}
	var out strings.Builder
	for _, fn := range ast.Functions {
		name := fn.Name
		if name == "" {
			name = "main"
		}
		out.WriteString("fun ")
		out.WriteString(name)
		out.WriteString("()")
		vars := map[string]bool{}
		for _, v := range ast.Vars {
			vars[v.Name] = false
		}
		lines := make([]string, len(fn.Lines))
		for i, ln := range fn.Lines {
			lines[i] = ln.Text
		}
		stmts, err := parseStatements(lines, vars, fn.StartLine)
		if err != nil {
			return nil, err
		}
		if len(stmts) == 0 {
			out.WriteString(" {}\n")
			continue
		}
		out.WriteString(" {\n")
		for _, st := range stmts {
			out.WriteString("  ")
			out.WriteString(st)
			out.WriteByte('\n')
		}
		out.WriteString("}\n")
	}
	if out.Len() == 0 {
		return nil, fmt.Errorf("no convertible functions found")
	}
	return []byte(out.String()), nil
}

// ConvertFile reads the COBOL file and converts it to Mochi.
func ConvertFile(path string) ([]byte, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return Convert(string(data))
}

func parseStatements(lines []string, vars map[string]bool, lineBase int) ([]string, error) {
	var out []string
	for i := 0; i < len(lines); i++ {
		ll := strings.TrimSpace(strings.TrimSuffix(lines[i], "."))
		switch {
		case strings.HasPrefix(ll, "DISPLAY "):
			expr := strings.TrimSpace(strings.TrimPrefix(ll, "DISPLAY "))
			expr = normalizeExpr(expr, vars)
			out = append(out, "print("+expr+")")

		case strings.HasPrefix(ll, "MOVE "):
			rest := strings.TrimSpace(strings.TrimPrefix(ll, "MOVE "))
			parts := strings.Split(rest, " TO ")
			if len(parts) == 2 {
				expr := normalizeExpr(strings.TrimSpace(parts[0]), vars)
				dest := strings.ToLower(strings.TrimSpace(parts[1]))
				dest = strings.ReplaceAll(dest, "(", "[")
				dest = strings.ReplaceAll(dest, ")", "]")
				baseVar := strings.Split(dest, "[")[0]
				if _, ok := vars[baseVar]; !ok {
					vars[baseVar] = true
				}
				if defined, ok := vars[dest]; ok && !defined {
					out = append(out, "var "+dest+" = "+expr)
					vars[dest] = true
				} else {
					if !ok {
						vars[dest] = true
					}
					out = append(out, dest+" = "+expr)
				}
			}

		case strings.HasPrefix(ll, "COMPUTE "):
			expr := strings.TrimSpace(strings.TrimPrefix(ll, "COMPUTE "))
			if strings.Contains(expr, "=") {
				parts := strings.SplitN(expr, "=", 2)
				dest := strings.ToLower(strings.TrimSpace(parts[0]))
				dest = strings.ReplaceAll(dest, "(", "[")
				dest = strings.ReplaceAll(dest, ")", "]")
				rhs := normalizeExpr(strings.TrimSpace(parts[1]), vars)
				baseVar := strings.Split(dest, "[")[0]
				if _, ok := vars[baseVar]; !ok {
					vars[baseVar] = true
				}
				if defined, ok := vars[dest]; ok && !defined {
					out = append(out, "var "+dest+" = "+rhs)
					vars[dest] = true
				} else {
					if !ok {
						vars[dest] = true
					}
					out = append(out, dest+" = "+rhs)
				}
			} else {
				out = append(out, expr)
			}

		case strings.HasPrefix(ll, "ADD "):
			rest := strings.TrimSpace(strings.TrimPrefix(ll, "ADD "))
			parts := strings.Split(rest, " TO ")
			if len(parts) == 2 {
				val := normalizeExpr(strings.TrimSpace(parts[0]), vars)
				dest := strings.ToLower(strings.TrimSpace(parts[1]))
				dest = strings.ReplaceAll(dest, "(", "[")
				dest = strings.ReplaceAll(dest, ")", "]")
				baseVar := strings.Split(dest, "[")[0]
				if _, ok := vars[baseVar]; !ok {
					vars[baseVar] = true
				}
				if _, ok := vars[dest]; !ok {
					vars[dest] = true
					out = append(out, "var "+dest+" = "+dest+" + "+val)
				} else {
					out = append(out, dest+" = "+dest+" + "+val)
				}
			}

		case strings.HasPrefix(ll, "PERFORM UNTIL "):
			cond := normalizeExpr(strings.TrimSpace(strings.TrimPrefix(ll, "PERFORM UNTIL ")), vars)
			body := []string{}
			for j := i + 1; j < len(lines); j++ {
				n := strings.TrimSpace(lines[j])
				if n == "END-PERFORM" {
					i = j
					break
				}
				body = append(body, lines[j])
			}
			isNot := strings.HasPrefix(cond, "NOT ")
			if isNot {
				cond = strings.TrimSpace(strings.TrimPrefix(cond, "NOT "))
				out = append(out, "while "+cond+" {")
			} else {
				out = append(out, "while !("+cond+") {")
			}
			inner, err := parseStatements(body, vars, lineBase+i+1)
			if err != nil {
				return nil, err
			}
			for _, st := range inner {
				out = append(out, "  "+st)
			}
			out = append(out, "}")

		case strings.HasPrefix(ll, "PERFORM VARYING "):
			rest := strings.TrimSpace(strings.TrimPrefix(ll, "PERFORM VARYING "))
			parts := strings.Split(rest, " FROM ")
			if len(parts) != 2 {
				break
			}
			varName := strings.TrimSpace(parts[0])
			vn := strings.ToLower(varName)
			if _, ok := vars[vn]; !ok {
				vars[vn] = true
			}
			rest = parts[1]
			parts = strings.Split(rest, " BY ")
			if len(parts) != 2 {
				break
			}
			startExpr := normalizeExpr(strings.TrimSpace(parts[0]), vars)
			rest = parts[1]
			parts = strings.Split(rest, " UNTIL ")
			if len(parts) != 2 {
				break
			}
			step := normalizeExpr(strings.TrimSpace(parts[0]), vars)
			cond := normalizeExpr(strings.TrimSpace(parts[1]), vars)
			body := []string{}
			for j := i + 1; j < len(lines); j++ {
				n := strings.TrimSpace(lines[j])
				if n == "END-PERFORM" {
					i = j
					break
				}
				body = append(body, lines[j])
			}
			end := ""
			if strings.Contains(cond, ">=") {
				end = normalizeExpr(strings.TrimSpace(strings.Split(cond, ">=")[1]), vars)
			} else if strings.Contains(cond, "<=") {
				end = normalizeExpr(strings.TrimSpace(strings.Split(cond, "<=")[1]), vars)
			}
			loop := "for " + strings.ToLower(varName) + " in range(" + startExpr + ", " + end
			if step != "1" && step != "-1" {
				loop += ", " + step
			}
			loop += ") {"
			out = append(out, loop)
			inner, err := parseStatements(body, vars, lineBase+i+1)
			if err != nil {
				return nil, err
			}
			for _, st := range inner {
				out = append(out, "  "+st)
			}
			out = append(out, "}")

		case strings.HasPrefix(ll, "IF "):
			cond := normalizeExpr(strings.TrimSpace(strings.TrimPrefix(ll, "IF ")), vars)
			thenLines := []string{}
			elseLines := []string{}
			j := i + 1
			inElse := false
			depth := 0
			for ; j < len(lines); j++ {
				n := strings.TrimSpace(lines[j])
				upper := strings.ToUpper(n)
				if strings.HasPrefix(upper, "IF ") {
					depth++
				}
				if upper == "END-IF" {
					if depth == 0 {
						break
					}
					depth--
					if inElse {
						elseLines = append(elseLines, lines[j])
					} else {
						thenLines = append(thenLines, lines[j])
					}
					continue
				}
				if depth == 0 && upper == "ELSE" {
					inElse = true
					continue
				}
				if inElse {
					elseLines = append(elseLines, lines[j])
				} else {
					thenLines = append(thenLines, lines[j])
				}
			}
			i = j
			out = append(out, "if "+cond+" {")
			inner, err := parseStatements(thenLines, vars, lineBase+i+1)
			if err != nil {
				return nil, err
			}
			for _, st := range inner {
				out = append(out, "  "+st)
			}
			if len(elseLines) > 0 {
				out = append(out, "} else {")
				inner2, err := parseStatements(elseLines, vars, lineBase+i+1)
				if err != nil {
					return nil, err
				}
				for _, st := range inner2 {
					out = append(out, "  "+st)
				}
			}
			out = append(out, "}")
		case ll == "" || ll == "EXIT" || ll == "EXIT PERFORM":
			// ignore
		case strings.HasPrefix(ll, "*>"):
			// comment
		case strings.HasPrefix(ll, "PERFORM "):
			name := strings.TrimSpace(strings.TrimPrefix(ll, "PERFORM "))
			name = strings.ToLower(strings.TrimSuffix(name, "."))
			out = append(out, name+"()")
		default:
			// skip unsupported statements but include context in error if needed
			continue
		}
	}
	return out, nil
}

func normalizeExpr(expr string, vars map[string]bool) string {
	for name := range vars {
		upper := strings.ToUpper(name)
		// convert array access like FOO(1) to foo[1]
		reIdx := regexp.MustCompile(`\b` + regexp.QuoteMeta(upper) + `\(([^)]+)\)`)
		expr = reIdx.ReplaceAllString(expr, name+`[$1]`)
	}
	for name := range vars {
		upper := strings.ToUpper(name)
		re := regexp.MustCompile(`\b` + regexp.QuoteMeta(upper) + `\b`)
		expr = re.ReplaceAllString(expr, name)
	}
	return expr
}

func snippet(lines []string, idx, start int) string {
	from := idx - 1
	if from < 0 {
		from = 0
	}
	to := idx + 1
	if to >= len(lines) {
		to = len(lines) - 1
	}
	var out []string
	for i := from; i <= to; i++ {
		out = append(out, fmt.Sprintf("%4d | %s", start+i, lines[i]))
	}
	return strings.Join(out, "\n")
}
