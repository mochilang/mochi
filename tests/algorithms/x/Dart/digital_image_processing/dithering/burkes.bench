{"duration_us":12419,"memory_bytes":4456448,"name":"_start"}