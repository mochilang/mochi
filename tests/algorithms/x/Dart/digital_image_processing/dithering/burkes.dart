// Generated by Mochi transpiler
import 'dart:convert';
import 'dart:io';

int _nowSeed = 0;
bool _nowSeeded = false;
void _initNow() {
  var s = Platform.environment['MOCHI_NOW_SEED'];
  if (s != null && s.isNotEmpty) {
    var v = int.tryParse(s);
    if (v != null) {
      _nowSeed = v;
      _nowSeeded = true;
    }
  }
}
int _now() {
  if (_nowSeeded) {
    _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647;
    return _nowSeed;
  }
  return DateTime.now().microsecondsSinceEpoch;
}

String _substr(String s, num start, num end) {
  var n = s.length;
  int s0 = start.toInt();
  int e0 = end.toInt();
  if (s0 < 0) s0 += n;
  if (e0 < 0) e0 += n;
  if (s0 < 0) s0 = 0;
  if (s0 > n) s0 = n;
  if (e0 < 0) e0 = 0;
  if (e0 > n) e0 = n;
  if (s0 > e0) s0 = e0;
  return s.substring(s0, e0);
}

int get_greyscale(int blue, int green, int red) {
  double b = blue as double;
  double g = green as double;
  double r = red as double;
  return 0.114 * b + 0.587 * g + 0.299 * r as int;
}

List<List<int>> zeros(int h, int w) {
  List<List<int>> table = <List<int>>[];
  int i = 0;
  while (i < h) {
    List<int> row = <int>[];
    int j = 0;
    while (j < w) {
    row = [...row, 0];
    j = j + 1;
  }
    table = ([...table, row] as List).map((e) => ((e as List).map((e) => (e is BigInt ? e.toInt() : (e as int))).toList() as List<int>)).toList();
    i = i + 1;
  }
  return table;
}

List<List<int>> burkes_dither(List<List<List<int>>> img, int threshold) {
  int height = img.length;
  int width = img[0].length;
  List<List<int>> error_table = zeros(height + 1, width + 4);
  List<List<int>> output = <List<int>>[];
  int y = 0;
  while (y < height) {
    List<int> row = <int>[];
    int x = 0;
    while (x < width) {
    List<int> px = img[y][x];
    int grey = get_greyscale(px[0], px[1], px[2]);
    int total = grey + error_table[y][x + 2];
    int new_val = 0;
    int current_error = 0;
    if (threshold > total) {
    new_val = 0;
    current_error = total;
  } else {
    new_val = 255;
    current_error = total - 255;
  }
    row = [...row, new_val];
    while (error_table[y]!.length <= x + 3) { error_table[y]!.add(0); } error_table[y]![x + 3] = error_table[y][x + 3] + 8 * current_error ~/ 32;
    while (error_table[y]!.length <= x + 4) { error_table[y]!.add(0); } error_table[y]![x + 4] = error_table[y][x + 4] + 4 * current_error ~/ 32;
    while (error_table[y + 1]!.length <= x + 2) { error_table[y + 1]!.add(0); } error_table[y + 1]![x + 2] = error_table[y + 1][x + 2] + 8 * current_error ~/ 32;
    while (error_table[y + 1]!.length <= x + 3) { error_table[y + 1]!.add(0); } error_table[y + 1]![x + 3] = error_table[y + 1][x + 3] + 4 * current_error ~/ 32;
    while (error_table[y + 1]!.length <= x + 4) { error_table[y + 1]!.add(0); } error_table[y + 1]![x + 4] = error_table[y + 1][x + 4] + 2 * current_error ~/ 32;
    while (error_table[y + 1]!.length <= x + 1) { error_table[y + 1]!.add(0); } error_table[y + 1]![x + 1] = error_table[y + 1][x + 1] + 4 * current_error ~/ 32;
    while (error_table[y + 1]!.length <= x) { error_table[y + 1]!.add(0); } error_table[y + 1]![x] = error_table[y + 1][x] + 2 * current_error ~/ 32;
    x = x + 1;
  }
    output = ([...output, row] as List).map((e) => ((e as List).map((e) => (e is BigInt ? e.toInt() : (e as int))).toList() as List<int>)).toList();
    y = y + 1;
  }
  return output;
}

void _main() {
  List<List<List<int>>> img = [[[0, 0, 0], [64, 64, 64], [128, 128, 128], [192, 192, 192]], [[255, 255, 255], [200, 200, 200], [150, 150, 150], [100, 100, 100]], [[30, 144, 255], [255, 0, 0], [0, 255, 0], [0, 0, 255]], [[50, 100, 150], [80, 160, 240], [70, 140, 210], [60, 120, 180]]];
  List<List<int>> result = burkes_dither(img, 128);
  int y = 0;
  while (y < result.length) {
    String line = "";
    int x = 0;
    while (x < result[y].length) {
    line = line + (result[y][x]).toString();
    if (x < result[y].length - 1) {
    line = line + " ";
  }
    x = x + 1;
  }
    print(line);
    y = y + 1;
  }
}

void _start() {
  var _benchMem0 = ProcessInfo.currentRss;
  var _benchSw = Stopwatch()..start();
  _initNow();
  {
  var _benchMem0 = ProcessInfo.currentRss;
  var _benchSw = Stopwatch()..start();
  _main();
  _benchSw.stop();
  var _benchMem1 = ProcessInfo.currentRss;
  print(jsonEncode({"duration_us": _benchSw.elapsedMicroseconds, "memory_bytes": (_benchMem1 - _benchMem0).abs(), "name": "main"}));
}
  _benchSw.stop();
  var _benchMem1 = ProcessInfo.currentRss;
  print(jsonEncode({"duration_us": _benchSw.elapsedMicroseconds, "memory_bytes": (_benchMem1 - _benchMem0).abs(), "name": "_start"}));
}

void main() => _start();
