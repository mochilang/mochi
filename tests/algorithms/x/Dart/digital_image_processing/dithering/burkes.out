Unhandled exception:
type 'int' is not a subtype of type 'double' in type cast
#0      get_greyscale (file:///workspace/mochi/tests/algorithms/x/Dart/digital_image_processing/dithering/burkes.dart:17:19)
#1      burkes_dither (file:///workspace/mochi/tests/algorithms/x/Dart/digital_image_processing/dithering/burkes.dart:50:16)
#2      _main (file:///workspace/mochi/tests/algorithms/x/Dart/digital_image_processing/dithering/burkes.dart:79:28)
#3      _start (file:///workspace/mochi/tests/algorithms/x/Dart/digital_image_processing/dithering/burkes.dart:97:3)
#4      main (file:///workspace/mochi/tests/algorithms/x/Dart/digital_image_processing/dithering/burkes.dart:100:16)
#5      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:314:19)
#6      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:193:12)